File: python_sdk_samples/schedule_samples/create_schedule.py
"""Create a Cadence schedule that fires ScheduleSampleWorkflow every minute.

Usage:
    uv run python -m schedule_samples.create_schedule
    uv run python -m schedule_samples.create_schedule --schedule-id my-other-schedule
"""

import argparse
import asyncio
import json
from datetime import timedelta

from google.protobuf.duration import from_timedelta

from cadence.api.v1 import common_pb2, schedule_pb2, tasklist_pb2
from cadence.client import Client
from cadence.error import CadenceRpcError

from schedule_samples.workflow import SCHEDULE_ID, TASK_LIST, WORKFLOW_TYPE


def _payload(value: object) -> common_pb2.Payload:
    # Cadence memo values are JSON-encoded as a single-element list, matching Go/Java SDK behavior.
    return common_pb2.Payload(data=json.dumps([value]).encode())


async def main(args: argparse.Namespace) -> None:
    schedule_id = args.schedule_id
    async with Client(domain=args.domain, target=args.target) as client:
        try:
            await client.create_schedule(
                schedule_id,
                spec=schedule_pb2.ScheduleSpec(cron_expression="* * * * *"),
                action=schedule_pb2.ScheduleAction(
                    start_workflow=schedule_pb2.ScheduleAction.StartWorkflowAction(
                        workflow_type=common_pb2.WorkflowType(name=WORKFLOW_TYPE),
                        task_list=tasklist_pb2.TaskList(name=TASK_LIST),
                        workflow_id_prefix=f"{schedule_id}-",
                        execution_start_to_close_timeout=from_timedelta(timedelta(minutes=10)),
                        task_start_to_close_timeout=from_timedelta(timedelta(seconds=10)),
                    )
                ),
                policies=schedule_pb2.SchedulePolicies(
                    overlap_policy=schedule_pb2.SCHEDULE_OVERLAP_POLICY_SKIP_NEW,
                    catch_up_policy=schedule_pb2.SCHEDULE_CATCH_UP_POLICY_SKIP,
                    pause_on_failure=True,
                ),
                memo=common_pb2.Memo(
                    fields={
                        "owner": _payload("platform-team"),
                        "env": _payload("dev"),
                    }
                ),
            )
            print(f"Created schedule {schedule_id!r} (fires every minute)")
        except CadenceRpcError as e:
            if "already exists" in str(e):
                print(f"Schedule {schedule_id!r} already exists. Run delete_schedule first or use --schedule-id to pick a different name.")
            else:
                raise


def build_parser() -> argparse.ArgumentParser:
    p = argparse.ArgumentParser(description="Create a Cadence schedule")
    p.add_argument("--target", default="localhost:7833", help="Cadence frontend host:port")
    p.add_argument("--domain", default="default", help="Cadence domain")
    p.add_argument("--schedule-id", default=SCHEDULE_ID, help="Schedule ID to create (default: %(default)s)")
    return p


if __name__ == "__main__":
    asyncio.run(main(build_parser().parse_args()))

File: new_samples/schedule/create_schedule.go
package main

import (
	"context"
	"fmt"
	"strings"
	"time"

	"go.uber.org/cadence/.gen/go/shared"
	"go.uber.org/cadence/client"
	"go.uber.org/zap"
)

func runCreate() {
	logger := BuildLogger()
	c := buildScheduleClient()

	action := startWorkflowAction(logger)
	action.WorkflowIDPrefix = ScheduleID + "-"

	_, err := c.Create(context.Background(), &client.CreateScheduleRequest{
		ScheduleID: ScheduleID,
		Spec:       &client.ScheduleSpec{CronExpression: "* * * * *"},
		Action:     &client.ScheduleAction{StartWorkflow: action},
		Policies: &client.SchedulePolicies{
			OverlapPolicy:  client.ScheduleOverlapPolicySkipNew,
			CatchUpPolicy:  client.ScheduleCatchUpPolicySkip,
			PauseOnFailure: true,
		},
		Memo: map[string]interface{}{
			"owner": "platform-team",
			"env":   "dev",
		},
	})
	if err != nil {
		if be, ok := err.(*shared.BadRequestError); ok && strings.Contains(be.Message, "already") {
			fmt.Printf("Schedule %q already exists. Run delete first or use a different schedule ID.\n", ScheduleID)
			return
		}
		logger.Fatal("Create failed", zap.Error(err))
	}
	fmt.Printf("Created schedule %q (fires every minute)\n", ScheduleID)

	desc, err := c.Describe(context.Background(), ScheduleID)
	if err != nil {
		logger.Fatal("Describe failed", zap.Error(err))
	}
	printSchedule(ScheduleID, desc)
}

func printSchedule(id string, desc *client.DescribeScheduleResponse) {
	fmt.Printf("\nSchedule : %q\n", id)
	if desc.Spec != nil {
		fmt.Printf("  cron           : %q\n", desc.Spec.CronExpression)
	}
	if desc.State != nil {
		fmt.Printf("  paused         : %v\n", desc.State.Paused)
		if desc.State.Paused && desc.State.PauseInfo != nil {
			fmt.Printf("  pause reason   : %q\n", desc.State.PauseInfo.Reason)
		}
	}
	if desc.Policies != nil {
		fmt.Printf("  overlap policy : %v\n", desc.Policies.OverlapPolicy)
		fmt.Printf("  pause_on_fail  : %v\n", desc.Policies.PauseOnFailure)
	}
	if desc.Info != nil {
		if !desc.Info.NextRunTime.IsZero() {
			fmt.Printf("  next run       : %s UTC\n", desc.Info.NextRunTime.UTC().Format(time.RFC3339))
		}
		if !desc.Info.LastRunTime.IsZero() {
			fmt.Printf("  last run       : %s UTC\n", desc.Info.LastRunTime.UTC().Format(time.RFC3339))
		}
		fmt.Printf("  total runs     : %d\n", desc.Info.TotalRuns)
	}
}
